fetch/api/abort/general.js

```diff
@@ -0,0 +1,288 @@
+}
+
+function fetch(request, {
+  signal = request.signal
+}={}) {
+  const url = request.url || request;
+
+  if (signal && signal.aborted) return Promise.reject(new DOMException('Aborted', 'AbortError'));
+
+  return new Promise((resolve, reject) => {
+    // Do the usual XHR stuff
+    const req = new XMLHttpRequest();
+    req.open('GET', url);
+
+    const read = new Promise((readResolve, readReject) => {
+      req.onload = () => {
+        readResolve();
+      };
+
+      req.onreadystatechange = () => {
+        if (req.readyState == XMLHttpRequest.HEADERS_RECEIVED) {
+          resolve({
+            async json() {
+              await read;
+              return JSON.parse(req.responseText);
+            },
+            async text() {
+              await read;
+              return req.responseText;
+            }
+          })
+        }
+      };
+
+      // Handle network errors
+      req.onerror = () => {
+        reject(Error("Network Error"));
+        readReject(Error("Network Error"));
+      };
+  
+      req.onabort = () => {
+        reject(new DOMException('Aborted', 'AbortError'));
+        readReject(new DOMException('Aborted', 'AbortError'));
+      };
+  
+      // Make the request
+      req.send();
+  
+      if (signal) {
+        signal._addAbortCallback(() => req.abort());
+      }
+    });
+  });
+}
+//*/
+
+function assert_abort_error(err) {
+  assert_equals(err.constructor, DOMException);
+  assert_equals(err.name, 'AbortError');
+}
+
+promise_test(async () => {
+  const controller = new AbortController();
+  const signal = controller.signal;
+  controller.abort();
+
+  await fetch('../resources/data.json', { signal }).then(() => {
+    assert_unreached("Fetch must not resolve");
+  }, err => {
+    // Using .catch rather than try/catch to ensure the promise
+    // is rejecting (rather than throwing).
+    assert_abort_error(err);
+  });
+}, "Aborting rejects with AbortError");
+
+promise_test(async () => {
+  const controller = new AbortController();
+  const signal = controller.signal;
+  controller.abort();
+
+  const log = [];
+
+  await Promise.all([
+    fetch('../resources/data.json', { signal }).then(
+      () => log.push('fetch-resolve'),
+      () => log.push('fetch-reject')
+    ),
+    Promise.resolve().then(() => log.push('next-microtask'))
+  ]);
+
+  assert_array_equals(log, ['fetch-reject', 'next-microtask']);
+}, "Already aborted signal rejects immediately");
+
+promise_test(async () => {
+  const controller = new AbortController();
+  const signal = controller.signal;
+  controller.abort();
+
+  const request = new Request('../resources/data.json', {
+    signal,
+    method: 'POST',
+    body: 'foo',
+    headers: { 'Content-Type': 'text/plain' }
+  });
+
+  await fetch(request).catch(() => {});
+
+  assert_false(request.bodyUsed, "Body has not been used");
+  assert_equals(await request.text(), 'foo', "Correct body");
+}, "Request is not 'used' if signal is aborted before fetching");
```

Comment:
This is where it gets messy I think. Here are my gut feelings:

```js
const request = new Request(url, { signal: abortedSignal, body });
const anotherRequest = new Request(request);
```

It feels like `request`'s body should be disturbed/locked, since the body is being explicitly moved from one request to another.

```js
fetch(invalidURL, { signal: abortedSignal });
```

The above should reject based on the invalid URL, not aborted.

```js
fetch(request, { signal: abortedSignal, body });
```

In the above, the body shouldn't be consumed, as it doesn't need to be sent to the server, as the signal is already aborted.

So, from the developer's point of view, the order is:

1. Validate the args.
1. Abort if already aborted.
1. Make the request.

Locking/disturbing the body as part of "Validate the args" doesn't fit well with me, although I realise that's how it's specced now.

I guess fetch could undisturb/unlock the body if it's already aborted?